sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java

```diff
@@ -961,33 +985,48 @@ public Flux<OpenConnectionResponse> openConnectionsAndInitCaches(
-                            .collect(Collectors.toList());
+                                    .filter(addressInfo -> this.protocolScheme.equals(addressInfo.getProtocolScheme()))
+                                    .collect(Collectors.groupingBy(Address::getParitionKeyRangeId))
+                                    .values()
+                                    .stream().map(addresses -> toPartitionAddressAndRange(collection.getResourceId(), addresses))
+                                    .collect(Collectors.toList());
 
                     return Flux.fromIterable(addressInfos)
-                            .flatMap(
-                                addressInfo -> {
-                                    this.serverPartitionAddressCache.set(addressInfo.getLeft(), addressInfo.getRight());
-
-                                    if (this.openConnectionsHandler != null) {
-                                        return this.openConnectionsHandler.openConnections(
-                                            collection.getResourceId(),
-                                            this.serviceEndpoint,
-                                            Arrays
-                                                .stream(addressInfo.getRight())
-                                                .map(addressInformation -> addressInformation.getPhysicalUri())
-                                                .collect(Collectors.toList()));
-                                    }
-
-                                    logger.info("OpenConnectionHandler is null, can not open connections");
-                                    return Flux.empty();
-                                },
-                                Configs.getCPUCnt() * 10,
-                                Configs.getCPUCnt() * 3);
+                            .flatMap(addressInfo -> {
+                                this.serverPartitionAddressCache.set(addressInfo.getLeft(), addressInfo.getRight());
+                                return Flux.fromArray(addressInfo.getRight());
+                            }, Configs.getCPUCnt() * 10, Configs.getCPUCnt() * 3)
+                            .flatMap(addressInformation -> Mono.just(new ImmutablePair<>(new ImmutablePair<>(containerLink, collection), addressInformation)));
                 });
+
+    }
+
+    public Flux<Void> submitOpenConnectionTask(
+            AddressInformation address,
+            DocumentCollection documentCollection,
+            int connectionsPerEndpointCount
+    ) {
+        if (this.openConnectionsHandler != null && this.proactiveOpenConnectionsProcessor != null) {
+
+            int connectionsRequiredForEndpoint = Math.max(connectionsPerEndpointCount,
+                    Configs.getMinConnectionPoolSizePerEndpoint());
+
+            OpenConnectionOperation openConnectionOperation = new OpenConnectionOperation(
```

Comment:
should `proactiveOpenConnectionsProcessor` has `openConnectionsHandler`? what is the benefits of containing a  openConnectionHandler for each openConnectionOperation?

Reply:
This is just a matter of consistency - `GatewayAddressCache` uses some instance of `openConnectionsHandler `to open a connection to some endpoint in the previous flow, in the same manner, an instance of `openConnectionHandler` is encapsulated in an `OpenConnectionOperation` along with the endpoint.

Reply:
should not all the open connections go throughput OpenConnectionsProcessor? instead some goes through openConnectionProcessor vs openConnectionHandler? 

And also currently openConnectionHandler does not have any concurrency checking, so by calling it directly will have a risk of opening connection too aggressive?